### ISSUE TITLE
feat: freeze daily experiment candidates

### ISSUE THREAD

**Issue by jmcclary27** [jmcclary27/Market-Regime-Detection-With-Auto-Model-Switching#37]
---
name: Freeze daily experiment candidates
about: Train, evaluate, package, and freeze the artifacts for the paper-trading study
title: "[Experiment] Freeze daily experiment candidates"
labels: "enhancement"
assignees: ""
---

## Problem and outcome

The frozen daily paper-trading experiment does not yet have one reproducible
entrypoint that turns the predeclared candidate set into a signed, immutable
experiment package. Model training, walk-forward evaluation, artifact
selection, Lambda bundling, and experiment metadata are currently separate
operations, which leaves room for mismatched model IDs, missing artifacts,
future-data leakage, accidental registry mutation, or an incomplete S3
reference.

Implement one command, for example:

```bash
python -m tools.freeze_experiment \
  --experiment-id frozen-daily-spy-v1 \
  --official-start-date 2026-09-01 \
  --data-cutoff 2026-08-31 \
  --features-path data/features/latest.parquet \
  --regimes-path data/regimes/latest.parquet \
  --output-dir artifacts/experiment/frozen-daily-spy-v1
```

The command must train/evaluate the fixed candidates, select the global Static
ML model, verify every required specialist, create a deterministic model
bundle, and write a complete manifest. It must not update the active registry,
published `latest` pointers, or any live deployment state.

## Acceptance criteria

- [ ] `tools.freeze_experiment` exposes a typed CLI with explicit experiment ID,
  official start date, data cutoff, input paths, output directory, random seed,
  and optional S3 publish arguments. It never infers the official start date
  from the current wall clock.

- [ ] The command creates the fixed candidate set using chronological,
  leakage-safe evaluation: global Ridge, global LightGBM, and one LightGBM
  specialist for each `bullish`, `sideways`, and `bearish` regime. Candidate
  artifacts remain under an experiment-specific candidate directory until the
  freeze succeeds.

- [ ] Static ML selection is deterministic and uses walk-forward net Sharpe as
  the primary ranking, then the smaller maximum-drawdown magnitude, then the
  greater cumulative return, then model ID as a stable final tie-breaker.
  The selected model ID and its scorecard row are recorded.

- [ ] The command verifies that the selected static model, all three regime
  specialists, the frozen HMM regime detector/state mapping, and the feature
  manifest exist, are readable, have valid model/artifact contracts, and have
  model IDs matching the prediction schema. Missing, candidate-only, invalid,
  non-finite, or regime-mismatched artifacts fail the command before any live
  pointer or manifest is published.

- [ ] The model bundle is created with deterministic member ordering, normalized
  archive metadata, and a bundle manifest containing every member path, size,
  and SHA-256. The bundle contains only the exact frozen artifacts needed by
  experiment inference, not unpublished candidate directories or unrelated
  runtime state.

- [ ] The local manifest records at least: schema version, experiment ID,
  official start date, data cutoff, creation time, symbols `(SPY, QQQ)`, traded
  symbol `SPY`, starting cash, fee/slippage assumptions, exposure thresholds,
  primary/secondary metrics, selected model IDs, model versions, artifact
  paths, artifact SHA-256 hashes, HMM/state-mapping references, feature-manifest
  reference, Git commit, bundle path/hash, and an explicit freeze status.

- [ ] `--publish-s3` uploads the bundle first and writes the manifest only after
  receiving and recording the bundle bucket, key, S3 VersionId, and SHA-256.
  The published manifest also records its own bucket/key/VersionId/SHA-256.
  S3 publishing requires explicit bucket/key arguments and fails closed when
  credentials, versioning, or returned VersionIds are unavailable.

- [ ] Re-running the command with the same inputs, seed, and output identity is
  byte-identical and succeeds idempotently. Re-running with changed inputs,
  dates, candidate IDs, or artifact hashes refuses to overwrite the frozen
  manifest or bundle.

- [ ] The command leaves `registry/active_model.yaml`, registry history,
  published model `latest` pointers, deployment events, and paper-account state
  unchanged. A failure leaves no partially published local freeze manifest and
  no published S3 manifest.

- [ ] The resulting manifest and bundle are accepted by the frozen experiment
  executor without manual model-ID/path edits. The existing active-plus-shadow
  inference path and legacy live simulator continue to pass their contract
  tests.

## Scope

- Likely source packages/files:
  - `tools/freeze_experiment.py` for the CLI/orchestration workflow.
  - `src/experiment/manifest.py` and `src/experiment/selection.py` for the
    manifest schema, candidate selection, and freeze validation.
  - `tools/train_lightgbm_expert.py`, `src/models/train.py`, and the existing
    walk-forward/evaluation helpers for candidate production and scorecards.
  - `tools/package_lambda_model_bundle.py` for deterministic bundle creation.
  - `tests/test_freeze_experiment.py` plus existing model, leakage, bundle, and
    Lambda contract tests.
  - `docs/frozen_daily_experiment.md` and the README command documentation.
- Data/model/artifact contracts affected:
  - New versioned freeze manifest schema and bundle manifest schema.
  - Candidate-to-frozen model identity and artifact hash contracts.
  - S3 bundle/manifest references containing bucket, key, VersionId, and SHA-256.
  - No changes to the legacy active registry schema or default pipeline output
    contracts.
- Documentation or CI changes required:
  - Document local freeze mode, S3 publish mode, required arguments, and the
    resulting artifact layout.
  - Add the focused freeze tests to the existing CI quality gate; no network or
    AWS credentials may be required for ordinary tests.
- Explicitly out of scope:
  - Daily Alpaca ingestion, EventBridge scheduling, Lambda deployment, and
    CloudFront/dashboard changes.
  - Starting or mutating the 90-day paper experiment.
  - Automatic retraining, promotion, rollback, or changes to
    `registry/active_model.yaml`.
  - 30/60/90-day reporting and final performance analysis.

## Safety and determinism

- Does this touch time-series ordering, target construction, regimes, model
  promotion, registry state, or trading simulation?

  Yes. It trains and evaluates time-series models and freezes regime-specific
  artifacts, but it must not promote a model or mutate the active registry.
  Training and selection must reuse chronological walk-forward splits and
  existing leakage guards. The data cutoff is an inclusive boundary: no row
  after it may enter training, regime fitting, scorecards, or the bundle.

- What prevents future-data leakage, nondeterminism, unsafe publication, or
  duplicate execution?

  The command sorts by timestamp and row identity, uses an explicit cutoff and
  seed, records the exact input hashes and Git commit, rejects non-finite or
  insufficient metrics, normalizes bundle ordering/metadata, writes local
  outputs atomically, and publishes the S3 manifest last. Candidate outputs are
  isolated from live discovery. A frozen identity cannot be overwritten with a
  different manifest or bundle.

- What is the failure behavior and rollback/disable path?

  Any failed training, evaluation, artifact-contract, hash, bundle, or S3
  versioning check exits non-zero and leaves the existing registry and frozen
  manifest unchanged. Local temporary files are cleaned up. If an S3 bundle has
  been uploaded but the manifest fails, it remains an unreferenced versioned
  object and is not live; disable the experiment by withholding/removing the
  manifest reference, never by rewriting the frozen object.

## Runtime and agentic context

- Is current evidence static, local runtime state, or production telemetry?

  Static repository code, deterministic test fixtures, local candidate artifacts,
  and optional authenticated S3 upload responses. No production telemetry is
  required or claimed.

- Does this require a future reviewer, verifier, MCP, or GitHub workflow?

  No connector, MCP, issue-to-agent workflow, or production reviewer is needed.
  A read-only review should trace each acceptance criterion to the manifest,
  bundle manifest, scorecard, and tests before any separate deployment issue is
  opened.

- If external access is needed, what is the minimum permission and data scope?

  Only `--publish-s3` needs AWS access: write permission to the designated
  versioned model-bundle prefix and designated manifest key, plus the ability
  to read the returned object VersionIds. No market-data, Lambda invocation,
  registry, delete, or broad-bucket permissions are needed.

## Validation plan

- Focused tests:
  - CLI smoke test using deterministic training fixtures and `tmp_path`.
  - Candidate/live separation: no active pointer or `latest` artifact changes.
  - Walk-forward cutoff and no-leakage tests, including a post-cutoff row that
    must not affect model selection or artifact hashes.
  - Static-model tie-break tests and required-regime coverage tests.
  - Missing/invalid/constant/non-finite artifact rejection tests.
  - Deterministic bundle byte/hash test and archive-traversal/layout validation.
  - Manifest schema, required-field, hash, freeze-idempotency, and immutable
    rewrite tests.
  - Fake-S3 publish test covering bundle-first ordering, VersionId capture,
    manifest-last behavior, missing-version failure, and retry idempotency.
  - Existing Lambda model-bundle and frozen-experiment contract tests.
- Full quality gate:
  - `ruff check .`
  - `ruff format --check .`
  - `mypy src tests`
  - `pytest -q`
- Integration/replay or manual evidence:
  - Run the command against the checked-in training fixture in local-only mode.
  - Inspect the generated selection scorecard, bundle manifest, and freeze
    manifest for cutoff, model IDs, hashes, and no active-registry mutation.
  - If S3 publish is explicitly authorized, run against a dedicated test
    prefix in a versioned bucket and retain the upload response/manifest as
    review evidence; do not use the production experiment prefix.
